src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -22,47 +31,132 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Diagnostics
-        ForegroundThreadAffinitizedObject,
-        ITaggerProvider
+    internal abstract partial class AbstractDiagnosticsTaggerProvider<TTag> : AsynchronousTaggerProvider<TTag>
         where TTag : ITag
     {
-        private readonly object _uniqueKey = new object();
         private readonly IDiagnosticService _diagnosticService;
-        private readonly IForegroundNotificationService _notificationService;
-        private readonly IAsynchronousOperationListener _listener;
 
         protected AbstractDiagnosticsTaggerProvider(
             IDiagnosticService diagnosticService,
             IForegroundNotificationService notificationService,
             IAsynchronousOperationListener listener)
+            : base(listener, notificationService)
         {
             _diagnosticService = diagnosticService;
-            _notificationService = notificationService;
-            _listener = listener;
         }
 
-        protected internal abstract IEnumerable<Option<bool>> Options { get; }
+        protected override TaggerDelay AddedTagNotificationDelay => TaggerDelay.OnIdle;
+
+        protected override ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            return TaggerEventSources.Compose(
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer, TaggerDelay.Medium),
+                TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer, TaggerDelay.Medium),
+                TaggerEventSources.OnDiagnosticsChanged(subjectBuffer, _diagnosticService, TaggerDelay.Short));
+        }
+
         protected internal abstract bool IsEnabled { get; }
         protected internal abstract bool IncludeDiagnostic(DiagnosticData data);
         protected internal abstract ITagSpan<TTag> CreateTagSpan(bool isLiveUpdate, SnapshotSpan span, DiagnosticData data);
 
-        ITagger<T> ITaggerProvider.CreateTagger<T>(ITextBuffer buffer)
+        protected override Task ProduceTagsAsync(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag, int? caretPosition)
         {
-            return CreateTagger<T>(buffer);
+            ProduceTags(context, spanToTag);
+            return SpecializedTasks.EmptyTask;
         }
 
-        public IAccurateTagger<T> CreateTagger<T>(ITextBuffer buffer) where T : ITag
+        private void ProduceTags(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag)
         {
-            var tagger = buffer.Properties.GetOrCreateSingletonProperty(
-                _uniqueKey, () => new AggregatingTagger(this, buffer));
-            tagger.OnTaggerCreated();
-            return tagger as IAccurateTagger<T>;
+            if (!this.IsEnabled)
+            {
+                return;
+            }
+
+            var document = spanToTag.Document;
+            if (document == null)
+            {
+                return;
+            }
+
+            if (!document.TryGetText(out var sourceText))
```

Comment:
Why is this preferred over just GetText()? It should be available either way so be equally fast. But if something goes sideways, this version is broken.

Reply:
agreed.  i can change to just using GetText.